Ingest the remaining entities and relationships

neurovelho opened the pull request:
This PR includes the remaining of the high value entities and relationships as discussed in the [issue thread](https://github.com/JupiterOne/graph-artifactory/issues/3).

Kindly let us know if something needs to be changed.

neurovelho commented:
Pinging @aiwilliams @ndowmon @austinkelleher

ndowmon commented on src/constants.ts:
```diff
@@ -59,10 +110,76 @@ export const relationships: Record<
+    _class: RelationshipClass.ALLOWS,
+    sourceType: entities.PERMISSION._type,
+    targetType: entities.REPOSITORY._type,
+  },
```
Actually, read comments in `/permissions.ts`. Not sure there should be _any_ `DENIES` relationships.

ndowmon commented on src/steps/permissions.ts:
```diff
@@ -0,0 +1,330 @@
+      relationships.PERMISSION_ALLOWS_BUILD,
+      relationships.PERMISSION_DENIES_BUILD,
+    ],
+    dependsOn: ['fetch-users', 'fetch-repositories', 'fetch-builds'],
```
should also depend on `fetch-groups`

ndowmon commented on src/steps/permissions.ts:
```diff
@@ -0,0 +1,330 @@
+    },
+  });
+
+  await jobState.addEntity(permissionEntity);
```
Looks like this code is adding two `artifactory_permission` entities for every permission. There should be just _one_ `artifactory_permission` entity created. That single entity should have relationships to both users _and_ groups.

neurovelho replied:
We were going by this comment here: https://github.com/JupiterOne/graph-artifactory/issues/3#issuecomment-678661547, but seeing as the response contains both users and groups, it would make more sense if we had both users and groups in the entity.

ndowmon replied:
Just to clarify - you mean both users and groups ~~in~~ _related to_ the entity, correct?

neurovelho replied:
Yes, that's right.

ndowmon commented on src/steps/permissions.ts:
```diff
@@ -0,0 +1,330 @@
+  );
+}
+
+function constructPermissionsMap(
```
Per artifactory [docs](https://www.jfrog.com/confluence/display/JFROG/Permissions), there can also be release bundle permissions, destination permissions, and pipeline permissions, PLUS "global permissions". Can you pull those additional permission targets in this function? I realize that some of those can't be used to target existing entities, but can you add code comments to implement others when they are implemented in the future?

neurovelho replied:
The api documentation seems to indicate that only repository, build and releaseBundle are available, though it could be that it's outdated :thinking: 

We'll definitely see if it's possible to add the rest.

ndowmon commented on src/steps/permissions.ts:
```diff
@@ -0,0 +1,330 @@
+  return targetPermissionsMap;
+}
+
+async function createPermissionBuildAllowsRelationships(
```
Jfrog docs discussing **include** and **exclude** patterns seem to indicate that both the statements are used to pattern match permission targets that are allowed by a given permission. I don't think the **exclude** pattern is actually a `DENIES` statement in reality.

I would instead suggest creating only `ALLOWS` statements for permissions targets that _do_ match the `include-pattern` and _do not_ match the `exclude-pattern`.

Also, I notice that artifactory exposes a pattern matching API - it would be better to use a matcher provided by the vendor than a nodejs utility. See https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API

```
Usage: GET /api/search/pattern?pattern=repo-key:this/is/a/*pattern*.war
```

ndowmon commented on src/types.ts:
```diff
@@ -95,18 +100,124 @@ export type ArtifactoryRepository = {
+      };
+      groups?: {
+        [name: string]: string[];
+      };
```
I haven't used the API, but it seems to me that `permission.repo` also has the `include-patterns` and `exclude-patterns` properties.

ndowmon commented on src/types.ts:
```diff
@@ -95,18 +100,124 @@ export type ArtifactoryRepository = {
+      };
     };
+    'include-patterns': string[];
+    'exclude-patterns': string[];
```
You should enumerate all of the permission target types, including `destinations`, `pipeline-sources`, `release-bundles`, and possibly `global` if it exists.

ndowmon reviewed (changes requested):
@neurovelho Thanks for all the work here! All the new entities and relationships look really good.

I peppered some requests throughout for you to review. In particular, I'd like you to take another look at the `artifactory_permissions` entities and relationships. Let me know if any of my comments don't make sense and we can discuss in more detail.

neurovelho commented on src/steps/access.ts:
```diff
@@ -136,6 +176,14 @@ export const accessSteps: IntegrationStep<IntegrationConfig>[] = [
+    id: 'fetch-access-tokens',
+    name: 'Fetch Access Tokens',
+    entities: [entities.ACCESS_TOKEN],
+    relationships: [relationships.ACCOUNT_HAS_ACCESS_TOKEN],
```
Looks like the admin can only create what's called admin access token, however it looks like it's actually possible to create regular access tokens that target a specific user using the rest api, so we'll change this.

neurovelho commented on src/steps/repositories.ts:
```diff
@@ -5,22 +5,37 @@ import {
+  return `artifactory_artifact:${uri}`;
+}
+
+function mapPackageTypeToStepEntityMetadata(type: string): StepEntityMetadata {
+  if (type.match(/docker|vagrant/i)) {
+    return entities.ARTIFACT_IMAGE;
+  }
+
+  return entities.ARTIFACT_CODEMODULE;
+}
+
```
Yeah I'm not too excited about this matcher either, I think it would be much better if we could just roll with one type.

ndowmon replied:
Spoke with @erichs about this. It really stinks that the API appears to only reliably return `folder` and `uri` properties, no additiona metadata: 
```
{
  "folder": false,
  "uri": "https://codeworkr.jfrog.io/artifactory/docker-repo/hello-world/1.1.1/sha256__bf756fb1ae65adf866bd8c456593cd24beb6a0a061dedf42b26a993176745f6b",
}
```

Since there are no other properties we can rely on, it doesn't seem like a great idea for us to add any custom matcher in our code (following the philosophy that we at J1 _only_ ingest the state of a target system, and don't build additional logic).

As long as you are pulling the `projectType` property on `artifactory_repository` entities, we can create J1QL like the following, which is good enough for now:

```
// look for maven builds
find 
  artifactory_repository that has 
  artifactory_artifact 
where 
  artifactory_repository.packageType = 'Maven'
```

neurovelho commented on src/steps/index.test.ts:
```diff
@@ -38,6 +46,11 @@ describe('JFrog Arrifactory', () => {
+    await fetchRepositories(context);
+    await fetchPipelineSources(context);
+    await fetchArtifacts(context);
+    await fetchBuilds(context);
```
Looks like we missed it, good catch.

neurovelho commented on src/client.ts:
```diff
@@ -161,16 +202,176 @@ export class APIClient {
   ): Promise<void> {
     const response = await this.request(
-      this.withBaseUri('artifactory/api/security/permissions'),
+      this.withBaseUri('artifactory/api/v2/security/permissions'),
```
These are the docs to the v2 api: https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API+V2, so it only contains a few extending endpoints. We used the v2 here because the response format looked a bit better.

neurovelho commented on src/steps/pipelineSources.ts:
```diff
@@ -0,0 +1,61 @@
+    id: 'fetch-pipeline-sources',
+    name: 'Fetch Pipeline Sources',
+    entities: [entities.PIPELINE_SOURCE],
+    relationships: [relationships.ACCOUNT_HAS_PIPELINE_SOURCE],
```
Well this is where things get interesting :smile:

So repositoryFullName here targets a git repo, in this case a just a dummy repo I created which contains some pipeline configuration. It's not related to an Artifactory repository.

Will definitely take a second look and see if there's a way to map this entity with other entities.

ndowmon replied:
Understood - seems more complex than I initially expected, if it ends up not being simple to implement a direct relationship you can disregard.

ndowmon commented on src/client.ts:
```diff
@@ -161,16 +202,176 @@ export class APIClient {
+    key: string,
+    iteratee: ResourceIteratee<ArtifactoryArtifactRef>,
+  ): Promise<void> {
+    if (['ANY', 'ANY LOCAL', 'ANY REMOTE'].includes(key)) {
```
Makes sense to me. If these are special, though, I would suggest using a different `_class` and `_type`. Probably `Group` and `artifactory_repository_group`, respectively. That way you won't have to filter them out when calling `iterateRepositoryArtifacts`.

neurovelho commented:
Alrighty, we've implemented the changes you requested @ndowmon.

There's however a few complications though:

- We implemented the user -> assigned -> access token relationship, however it looks like no matter what you do in the dashboard or API, the access tokens simply won't show up in the api/token response -- only the admin tokens do. Sadly this seems like a regular occurrence with this API.
- After trying out multiple things it seems like API only returns repository, build and releaseBundle permissions:

  **Release bundle** permission is difficult/impossible to implement because they aren’t available during the trial.

  **Pipeline permissions** can’t seem to be fetched even after we create them in the permission using JFrog admin dashboard, the API simply isn’t returning those for some reason.

  **Destination permissions** - we aren’t exactly sure what these represent, we don’t see additional field(s) for setting this on the permission setting page nor does the API returns anything about them.
- The pattern matching endpoint can only be used for artifacts, nothing else. That means that when we're matching include/exclude patterns for repositories and builds we pretty much have to use the external library for matching.

ndowmon commented on src/steps/permissions.ts:
```diff
@@ -101,40 +138,22 @@ async function createPermissionBuildAllowsRelationships(
+      _class: RelationshipClass.ALLOWS,
+      _type: relationships.PERMISSION_ALLOWS_REPOSITORY._type,
     },
+    (buildEntity) => `${permissionEntity._key}|allows|${buildEntity._key}`,
```
Not a big deal since it's functionally correct, but here it should be `permissionEntity`.

neurovelho replied:
Did you mean `repositoryEntity`? In any case, I renamed it since it was bothering me too much :slightly_smiling_face:.

ndowmon replied:
Yes! Thanks.

ndowmon commented on src/steps/permissions.ts:
```diff
@@ -237,9 +272,11 @@ export const permissionsSteps: IntegrationStep<IntegrationConfig>[] = [
       relationships.PERMISSION_ASSIGNED_GROUP,
       relationships.PERMISSION_ALLOWS_REPOSITORY,
       relationships.PERMISSION_ALLOWS_BUILD,
+      relationships.PERMISSION_ALLOWS_REPOSITORY_GROUP,
```
Cool, thanks for adding this!

ndowmon commented on src/steps/index.test.ts:
```diff
@@ -51,6 +52,7 @@ describe('JFrog Arrifactory', () => {
     await fetchPipelineSources(context);
     await fetchArtifacts(context);
     await fetchBuilds(context);
+    await fetchPermissions(context);
```
Thanks for adding. Not necessary now, but let's add a GH issue to add an `expect` statement for permissions here.

ndowmon reviewed (approved):
Thanks for making the requested changes! This integration definitely looks ready for deployment to me.

I commented a few small requests but we can easily open new github issues for those and fix in the future.

ndowmon reviewed (approved):
Thanks @neurovelho. This is a great integration!
